File: polytopes/build_functions/build_root_polytope_A3.js
// Root Polytope A3
// File: polytopes/builder_functions/build_root_polytope_A3.js
import { Polytope } from '../Polytope.js';

export function build_root_polytope_A3() {
  const roots4 = [];
  // ±(e_i - e_j)
  for (let i = 0; i < 4; i++) for (let j = i + 1; j < 4; j++) {
    const v = [0, 0, 0, 0]; v[i] = 1; v[j] = -1;
    roots4.push(v); roots4.push(v.map(x => -x));
  }
  const dot = (u, v) => u.reduce((s, ui, k) => s + ui * v[k], 0);
  const sub = (u, v) => u.map((ui, k) => ui - v[k]);
  const scale = (u, s) => u.map(ui => ui * s);
  const norm = u => Math.hypot(...u);
  const orthonormalBasis = vs => {
    const es = [];
    vs.forEach(v => {
      let u = [...v];
      es.forEach(e => { const p = dot(u, e); u = sub(u, scale(e, p)); });
      const n = norm(u);
      if (n > 1e-8) es.push(scale(u, 1 / n));
    });
    return es;
  };
  const basis4 = [
    [1, -1, 0, 0], [0, 1, -1, 0], [0, 0, 1, -1]
  ];
  const ons = orthonormalBasis(basis4);
  const verts3 = roots4.map(v4 => ons.map(e => dot(v4, e)));
  return new Polytope(verts3);
}

File: polytopes_old/build_functions/build_root_polytope_A3.py
#!/usr/bin/env sage
# File: build_root_polytope_A3.py


import math
try:
    from sage.all import Polyhedron
except ImportError:
    Polyhedron = None

def _placeholder_polytope():
    return None, "Error: Sage Not Loaded"

def build_root_polytope_A3():
    """
    Convex hull of roots ±(e_i−e_j), 1≤i<j≤4, projected into R³.
    """
    roots = []
    for i in range(4):
        for j in range(i+1,4):
            v=[0]*4; v[i]=1; v[j]=-1
            roots.append(v); roots.append([-x for x in v])

    # basis
    def dot(u,v): return sum(x*y for x,y in zip(u,v))
    def sub(u,v): return [x-y for x,y in zip(u,v)]
    def scale(u,s): return [x*s for x in u]
    def norm(u): return math.sqrt(dot(u,u))
    raw=[[1,-1,0,0],[0,1,-1,0],[0,0,1,-1]]
    ons=[]
    for b in raw:
        u=b[:]
        for e in ons:
            p=dot(u,e); u=sub(u,scale(e,p))
        nrm=norm(u)
        if nrm>1e-8: ons.append(scale(u,1/nrm))

    verts3=[tuple(dot(v,e) for e in ons) for v in roots]

    if not Polyhedron: return _placeholder_polytope()
    poly=Polyhedron(vertices=verts3)
    try:
        if poly.dim()>3: poly=poly.projection(0,1,2)
    except: pass
    return poly, "RootPolytope A3"
